csharp/src/Apache.Arrow/Ipc/ArrowStreamWriter.cs

```diff
@@ -45,17 +45,19 @@ internal class ArrowRecordBatchFlatBufferBuilder :
             IArrowArrayVisitor<StringArray>,
             IArrowArrayVisitor<BinaryArray>
         {
-            public struct Buffer
+            public readonly struct Buffer
             {
                 public readonly ArrowBuffer DataBuffer;
                 public readonly int Offset;
                 public readonly int Length;
+                public readonly int Padding;
```

Comment:
Does this actually need to be stored in Buffer?  Seems like the call site in WriteRecordBatchInternalAsync could just compute this value instead?

Reply:
It could be computed in WriteRecordBatchInternalAsync , but then computing the padding (i.e. calling `RoundUpToMultipleOf8`) would need to be in both places.
It is needed when creating Buffer because we need to increment the `TotalLength` to include the padding. That way the next Buffer's Offset is always aligned to 8 bytes.

---
In reply to: [302357499](https://github.com/apache/arrow/pull/4851#discussion_r302357499) [](ancestors = 302357499)

Reply:
Yes, but that computation is cheap, and it looks like this increases the size of the buffer, which could be more expensive overall.  Something to think about.

Reply:
I added a quick benchmark for ArrowStreamWriter, and noticed that I can remove ~10% of the allocated memory during Write by making this struct smaller - I could also remove the `Length` field. The running time is not affected. So I went forward with the change, thanks for the suggestion.

## Original proposal

|     Method | BatchLength | ColumnSetCount |         Mean |       Error |     StdDev | Gen 0/1k Op | Allocated Memory/Op |
|----------- |------------ |--------------- |-------------:|------------:|-----------:|------------:|--------------------:|
| WriteBatch |       10000 |             10 |     832.5 us |    52.00 us |   151.7 us |           - |            44.38 KB |
| WriteBatch |       10000 |             25 |   1,776.1 us |    61.54 us |   178.5 us |           - |           121.09 KB |
| WriteBatch |     1000000 |             10 |  61,126.2 us | 1,162.39 us | 1,141.6 us |           - |            44.38 KB |
| WriteBatch |     1000000 |             25 | 150,407.0 us | 2,938.05 us | 3,383.5 us |           - |           121.09 KB |

## With new changes

|     Method | BatchLength | ColumnSetCount |         Mean |       Error |     StdDev | Gen 0/1k Op | Allocated Memory/Op |
|----------- |------------ |--------------- |-------------:|------------:|-----------:|------------:|--------------------:|
| WriteBatch |       10000 |             10 |     827.4 us |    47.40 us |   138.3 us |           - |            40.41 KB |
| WriteBatch |       10000 |             25 |   1,758.5 us |    60.13 us |   176.3 us |           - |           105.13 KB |
| WriteBatch |     1000000 |             10 |  61,222.3 us | 1,391.92 us | 1,429.4 us |           - |            40.41 KB |
| WriteBatch |     1000000 |             25 | 149,516.4 us | 1,198.86 us |   936.0 us |           - |           105.13 KB |